packages/aws_common/lib/src/result/aws_result.dart

```diff
@@ -10,69 +10,141 @@ import 'package:aws_common/aws_common.dart';
 /// throw an exception if an exception occurred. See [exception] for more
 /// details.
 /// {@endtemplate}
-class AWSResult<T, E extends Exception>
-    with AWSEquatable<AWSResult<T, E>>, AWSDebuggable {
+sealed class AWSResult<V extends Object?, E extends Exception>
+    with AWSDebuggable, AWSSerializable<Map<String, Object?>> {
   /// Creates a failed result.
-  const AWSResult.error(E this.exception, [this.stackTrace])
-      : type = AWSResultType.error,
-        _value = null;
+  const factory AWSResult.error(E exception, [StackTrace? stackTrace]) =
+      AWSErrorResult<V, E>;
 
   /// Creates a successful result.
-  const AWSResult.success(T value)
-      : _value = value,
-        type = AWSResultType.success,
-        exception = null,
-        stackTrace = null;
+  const factory AWSResult.success(V value) = AWSSuccessResult<V, E>;
 
-  /// The value of the result, or null.
-  final T? _value;
+  const AWSResult._();
 
   /// The exception that occurred while attempting to retrieve the value.
-  final E? exception;
+  E? get exception;
 
   /// The original stack trace of [exception], if provided.
-  final StackTrace? stackTrace;
+  StackTrace? get stackTrace;
 
   /// Indicates if the result was a success.
-  final AWSResultType type;
+  @Deprecated('Use pattern matching instead')
+  AWSResultType get type;
 
   /// The value of the result, if the result was successful.
   ///
   /// If an exception was thrown while retrieving the value, this will throw.
   /// See [exception] for more details.
-  T get value {
-    switch (type) {
-      case AWSResultType.success:
-        // value will be non-null since it is required in AWSResult.success.
-        return _value!;
-      case AWSResultType.error:
-        if (stackTrace != null) {
-          // TODO(dnys1): Chain, instead, so that the current stack trace can
-          /// provide context to the original
-        }
-        // ignore: only_throw_errors
-        throw exception!;
-    }
-  }
+  V get value;
 
   /// The value of the result, or null if there was an error retrieving it.
-  T? get valueOrNull {
-    switch (type) {
-      case AWSResultType.success:
-        // value will be non-null since it is required in AWSResult.success.
-        return _value!;
-      case AWSResultType.error:
-        return null;
+  V? get valueOrNull;
```

Comment:
given `V extends Object?` I'm not sure if we need this

Reply:
The `value` getter throws for error results whereas this returns `null`